### PR TITLE
ADAL -> MSAL Migration for ARO-Installer

### PR DESCRIPTION
**Pull request by safwank97**
**Which issue this PR addresses:**

[[ARO-2135] AAD: Migrate ARO-Installer ADAL -> MSAL](https://issues.redhat.com/browse/ARO-2135)

**What this PR does / why we need it:**

In the end of the last fiscal year Microsoft [announced](https://techcommunity.microsoft.com/t5/microsoft-entra-azure-ad-blog/azure-ad-change-management-simplified/ba-p/2967456) the impending end-of-support for Azure Active Directory Authentication Library (ADAL): 

>  Update December 15th, 2022: ADAL end of support is now extended to June 30th, 2023. We will retire AAD Graph API any time after June 30th, 2023. Through the next six months (January 2023 – June 2023) we will continue informing customers about the upcoming end of support along with providing guidance on migration.

Since ARO-Installer is the subset of ARO-RP most of the code changes are mimicking the RP we can understand it with a reference PR -> [Migrate RP from adal to azidentity package #2667](https://github.com/Azure/ARO-RP/pull/2667), except for the code cleanup part which we can observe through this PR, a short example can be that removal of Refreshable Library will not be the blocker in case of ARO-Installer.

We need to migrate the entire Installer code to Microsoft Authentication Library (MSAL) which for us means using the [Azure Identity Client Module for Go or "azidentity"](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azidentity) in place of [Azure Active Directory authentication for Go or "adal"](https://pkg.go.dev/github.com/Azure/go-autorest/autorest/adal).

> Azure Identity or "azidentity" module provides token authentication support across the Azure SDK, including set of TokenCredential implementations, which can be particularly useful for clients supporting token authentication.
Whereas, the Azure Active Direcrory authentication for Go or "adal" was a standalone package for authentication.

(Note: "azidentity", despite its slightly outdated [README](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azidentity/README.md), [does now use MSAL](https://github.com/Azure/azure-sdk-for-go/issues/14920).)

**Test plan for issue:**

**Is there any documentation that needs to be updated for this PR?**

None that I'm aware of. As mentioned earlier Installer code base is subset of RP code base and with the referenced PR [#2667](https://github.com/Azure/ARO-RP/pull/2667) above it was an internal change to the RP and was not altering it's behavior so the same way the change should not alter Installer's behavior as well.

**Additional notes:**

- This does bump the Go requirements to 1.18 since [azidentity requires Go 1.18 as of v1.0](https://github.com/Azure/azure-sdk-for-go/commit/9514e5f7779d3406fa8320777da06867b79dfcc1#diff-5f1a0df6b303a95cf03dc9fb99c9bef2ddaddbcb963ffe6a95900ae263169791).
- Trying to break most of the commits up by the source directory.
- The first commit mostly includes new vendor files and modifications in old vendor files go.mod and go.sum files.
- azidentity [handles token refreshing automatically](https://github.com/Azure/azure-sdk-for-go/issues/10567#issuecomment-921376485), so we should no longer have to worry about this in the Installer. Therefore, this Pull Request eliminates [AuthorizationRefreshingAction](https://github.com/Azure/ARO-RP/blob/master/pkg/util/steps/refreshing.go) and [refreshable.Authorizer](https://github.com/Azure/ARO-RP/blob/master/pkg/util/refreshable/refreshable.go).
- Followup work: Per the [azidentity Migration Guide](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azidentity/MIGRATION.md#use-azidentity-credentials-with-older-packages), this PR vendors in an unofficial module https://github.com/jongio/azidext to serve as a temprary shim between azidentity's TokenCredential APIs and the older Azure SDK Client packages under https://github.com/Azure/go-autorest.

### CODE REVIEW

**Review by SudoBrendan — CHANGES_REQUESTED** [review]
One comment. Otherwise, there's still one `import` of the `adal` library that needs to be addressed in `pkg/util/refreshable/refreshable.go`. See Amber's PR for help on a migration path: https://github.com/Azure/ARO-RP/pull/2890

Eventually as an end state, you should be able to run `go mod tidy` then `go mod vendor` and not see `github.com/Azure/go-autorest/autorest/adal` show up in either `go.mod` or in `/vendor`.

Nearly there, great work!

File: pkg/util/instancemetadata/prod.go
Comment (safwank97) [reply pkg/util/instancemetadata/prod.go:21]: Yes I'll include these changes along with my next upcoming commits

**Review by mbarnes — CHANGES_REQUESTED** [review]
This is looking good so far.  In addition to what @SudoBrendan mentioned, the next thing I'd like to see is `pkg/env/armhelper.go` removed.  There's a few places under `pkg/env` where `ARMHelper` is declared in a struct or instantiated, but it's not used in ARO-Installer.

So...
```
$ git rm pkg/env/armhelper.go
$ make
fix the compiler errors by removing references to ARMHelper
$ make
until compilation is successful
$ git commit -a
```

Then, in a separate commit, you can just drop the `graphrbac` Azure client:
```
$ git rm -r pkg/util/azureclient/graphrbac
$ make
$ git commit -a
```
That's all you need to do to complete [ARO-2136: AAD: Migrate ARO-Installer Azure AD Graph -> Microsoft Graph](https://issues.redhat.com/browse/ARO-2136).  After you remove `armhelper.go` there's no AD Graph calls left to migrate.

File: Dockerfile.aro
Comment (mbarnes) [inline Dockerfile.aro:10]: This is from https://github.com/openshift/ARO-Installer/pull/31.

You might want to `git pull --rebase upstream main` on your local branch to pick this commit up from **main**, then `git push --force origin HEAD` so it's not repeated here.

File: cmd/aro/create.go
Comment (mbarnes) [inline cmd/aro/create.go:171]: I thought we were going to revert this change?  (cc: @SudoBrendan)

File: pkg/env/msiauthorizer.go
Comment (mbarnes) [inline pkg/env/msiauthorizer.go:20]: I think we can delete this type, and the `msiContext` parameter on `NewMSIAuthorizer`.  If there's only one option then it's pointless.

**Review by mbarnes — APPROVED** [review]
LGTM.  Left a couple suggestions but nothing worth blocking the PR over.  Could be done as followups.

File: cmd/aro/create.go
Comment (tony-schndr) [reply cmd/aro/create.go:171]: I think this is OK to leave.  Originally `FPAuthorizer` returned a `refreshable` now we call `refreshable.NewAuthorizer` directly which uses `FPAuthorizer` underneath.  When trying to revert this it opened a can of worms since the new version of `FPAuthorizer` returns an `autorest.Authorizer` instead of a `refreshable.Authorizer`.

Comment (safwank97) [reply cmd/aro/create.go:171]: Yes indeed when we reverted it back many things were breaking.

File: pkg/util/azureclient/environments.go
Comment (SudoBrendan) [inline pkg/util/azureclient/environments.go:53]: /!\ This needs to be changed to the USGov endpoints:

```
              ResourceManagerScope:      azure.USGovernmentCloud.ResourceManagerEndpoint + "/.default",
              KeyVaultScope:             azure.USGovernmentCloud.ResourceIdentifiers.KeyVault + "/.default",
              ActiveDirectoryGraphScope: azure.USGovernmentCloud.GraphEndpoint + "/.default",
```

File: pkg/util/steps/refreshing.go
Comment (SudoBrendan) [inline pkg/util/steps/refreshing.go:76]: nit: Unsure if this is in scope, but there's a drift here [between this repo and ARO-RP](https://github.com/Azure/ARO-RP/blob/master/pkg/util/steps/refreshing.go#L71-L78), which also tracks `azureerrors.IsInvalidSecretError(err)` - should we follow suit here?

File: pkg/util/steps/runner_test.go
Comment (SudoBrendan) [inline pkg/util/steps/runner_test.go:33]: nit: If this is unused, we should remove it instead of commenting.

**Review by SudoBrendan — CHANGES_REQUESTED** [review]
I did a review the lazy way: a `git diff` between this branch and `ARO-RP/master`. There's one critical diff that absolutely must be addressed before merge: make sure in USGov we are using the USGov endpoints. There are a few other diffs I discovered that could be addressed or not. Great work so far.

File: pkg/env/msiauthorizer.go
Comment (SudoBrendan) [reply pkg/env/msiauthorizer.go:20]: I think this is OK to do in a follow-up PR. We can throw those in with the RHCOS changes.

**Review by SudoBrendan — APPROVED** [review]
LGTM, the diff is now clean with the RP. We can follow-up on the /hack use of adal and a few other comments in this PR in a subsequent PR, since those will not affect cluster installation.

**Pull request merged**
